src/engine/guidance/post_processing.cpp

```diff
@@ -292,6 +296,130 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
+
+        // and update the leg geometry indices for the removed entry
+        std::for_each(steps.begin(), steps.end(), [](RouteStep &step)
+                      {
+                          --step.geometry_begin;
+                          --step.geometry_end;
+                      });
+    }
+
+    // make sure we still have enough segments
+    if (steps.size() <= 2)
+        return;
+
+    auto &next_to_last_step = *(steps.end() - 2);
+    // in the end, the situation with the roundabout cannot occur. As a result, we can remove all
+    // zero-length instructions
+    if (next_to_last_step.distance <= std::numeric_limits<double>::epsilon())
+    {
+        geometry.locations.pop_back();
+        geometry.segment_offsets.pop_back();
+        BOOST_ASSERT(geometry.segment_distances.back() < std::numeric_limits<double>::epsilon());
+        geometry.segment_distances.pop_back();
+
+        next_to_last_step.maneuver = detail::stepManeuverFromGeometry(
+            TurnInstruction::NO_TURN(), WaypointType::Arrive, geometry);
+        steps.pop_back();
+        // the geometry indices of the last step are already correct;
+    }
+}
+
+// assign relative locations to depart/arrive instructions
+std::vector<RouteStep> assignRelativeLocations(std::vector<RouteStep> steps,
+                                               const LegGeometry &leg_geometry,
+                                               const PhantomNode &source_node,
+                                               const PhantomNode &target_node)
+{
+    // We report the relative position of source/target to the road only within a range that is
+    // sufficiently different but not full of the path
+    const constexpr double MINIMAL_RELATIVE_DISTANCE = 5., MAXIMAL_RELATIVE_DISTANCE = 300.;
+    const auto distance_to_start = util::coordinate_calculation::haversineDistance(
+        source_node.input_location, leg_geometry.locations[0]);
+    const auto initial_modifier =
+        distance_to_start >= MINIMAL_RELATIVE_DISTANCE &&
+                distance_to_start <= MAXIMAL_RELATIVE_DISTANCE
+            ? angleToDirectionModifier(util::coordinate_calculation::computeAngle(
+                  source_node.input_location, leg_geometry.locations[0], leg_geometry.locations[1]))
+            : extractor::guidance::DirectionModifier::UTurn;
+
+    steps.front().maneuver.instruction.direction_modifier = initial_modifier;
+
+    const auto distance_from_end = util::coordinate_calculation::haversineDistance(
+        target_node.input_location, leg_geometry.locations.back());
+    const auto final_modifier =
+        distance_from_end >= MINIMAL_RELATIVE_DISTANCE &&
+                distance_from_end <= MAXIMAL_RELATIVE_DISTANCE
+            ? angleToDirectionModifier(util::coordinate_calculation::computeAngle(
+                  leg_geometry.locations[leg_geometry.locations.size() - 2],
+                  leg_geometry.locations[leg_geometry.locations.size() - 1],
```

Comment:
Note: If you're fine with getting an exception (instead of UB) you can use `.at()` which always bounds-checks instead of the subscript operator `[]`.